File: python-packages/core/src/tsv_data_analytics/file_paths_data_reader.py
"""FilePathsDataReader class"""

import sys

# local import
from tsv_data_analytics import file_paths_reader
from tsv_data_analytics import file_paths_util

class FilePathsDataReader:
    """A simple class to read filepaths data"""

    s3_region = ""
    aws_profile = ""  
 
    # file paths reader is an iterator based reader 
    file_paths_readers = None 

    # cur_data holds the current data block
    cur_data = None

    # cur_index is the pointer to the line in cur_data
    cur_index = 0
 
    # header is the header line for entire dataset
    header = None
   
    # constructor 
    def __init__(self, filepaths, s3_region, aws_profile):
        # s3 config
        self.s3_region = s3_region
        self.aws_profile = aws_profile

        # initialize the reader
        self.file_paths_readers = file_paths_reader.FilePathsReader(filepaths)

        # loop through all files until find a one with proper data
        found = False
        while (found == False):
            if (self.file_paths_readers.has_next()):
                # initial load
                filepath = self.file_paths_readers.next()
                self.cur_data = file_paths_util.read_file_content_as_lines(filepath, self.s3_region, self.aws_profile)

                # check for validity
                if (len(self.cur_data) == 0):
                    print("FilePathsDataReader: Invalid file found. No header.", filepath)
                    sys.exit(0)

                # read header
                self.header = self.cur_data[0].rstrip("\n")

                # assign cur_index to 0 for reading header
                self.cur_index = 1

                # check if found a valid file
                if (self.cur_index < len(self.cur_data)):
                    found = True
            else:
                self.cur_data = None
                self.cur_index = -1
                break

        # check if found some valid file to read
        if (found == False):
            self.cur_data = None
            self.cur_index = -1

    # get header
    def get_header(self):
        return self.header

    # has next returns boolean if there is still some data left
    def has_next(self):
        # check if no data block loaded
        if (self.cur_data is None):
            return False

        return True

    # close the file reader
    def close(self):
        self.cur_data = None
        self.cur_index = -1

    # next return the next line and also moves the pointers
    def next(self):
        # declare and initialize variable
        result = None

        # cur_data should be non empty
        if (self.has_next() == False):
            print("FilePathsDataReader: next() has next is false and next is called")
            return None
        
        # check for current pointer
        if (self.cur_index < len(self.cur_data)):
            result = self.cur_data[self.cur_index].rstrip("\n")
            self.cur_index = self.cur_index + 1
        else:
            print("FilePathsDataReader: next() this should not happen")

        # Check if cur_index has reached the end of data block
        if (self.cur_index >= len(self.cur_data)):
            # check if any more data blocks are left
            if (self.file_paths_readers.has_next() == False):
                self.cur_data = None
                self.cur_index = -1
            else:
                found = False
                while (found == False):
                    if (self.file_paths_readers.has_next()):
                        # read the next block and reinitialize the cur_index
                        filepath = self.file_paths_readers.next()
                        self.cur_data = file_paths_util.read_file_content_as_lines(filepath, self.s3_region, self.aws_profile)
                        self.cur_index = 1

                        # check if found a valid file
                        if (self.cur_index < len(self.cur_data)):
                            found = True 
                    else:
                        # invalidate the buffers
                        self.cur_data = None
                        self.cur_index = -1
                        break

        # return result
        return result

File: projects/lokki/scala/src/main/scala/omigo_core/FilePathsDataReader.scala
package omigo_core

class FilePathsDataReader(val filePaths: List[String], val s3Region: String, val awsProfile: String) {
  var filePathsReaders = new FilePathsReader(filePaths) 
  var curData: List[String] = null
  var curIndex: Int = 0
  var header: String = null

  def apply() {
    var found = false
    var flag = true
    while (flag == true && found == false) {
      if (filePathsReaders.hasNext()) {
        // initial load
        val filePath = filePathsReaders.next()
        curData = FilePathsUtil.read_file_content_as_lines(filePath, s3Region, awsProfile)

        // check for validity
        if (curData.length == 0) {
          // TODO: change this in python
          throw new Exception("FilePathsDataReader: Invalid file found. No header." + filePath)
        } 
        
        // read header
        header = curData(0).stripSuffix("\n")

        // assign cur_index to 0 for reading header
        curIndex = 1

        // check if found a valid file
        if (curIndex < curData.length)
          found = true
      } else {
        curData = null 
        curIndex = -1
        flag = false
      }
    }

    // check if found some valid file to read
    if (found == false) {
      curData = null 
      curIndex = -1
    }
  }

  // get header
  def getHeader() = {
      header
  }

  // has next returns boolean if there is still some data left
  def hasNext() = {
    // check if no data block loaded
    curData != null
  }

  // close the file reader
  def close(): Unit = {
    curData = null 
    curIndex = -1
  }

  // next return the next line and also moves the pointers
  def next(): String = {
    // declare and initialize variable
    var result: String = null 

    // cur_data should be non empty
    if (hasNext() == false) {
        // TODO: change python code
        throw new Exception("FilePathsDataReader: next() has next is false and next is called")
    }

    // check for current pointer
    if (curIndex < curData.length) {
        result = curData(curIndex).stripSuffix("\n")
        curIndex = curIndex + 1
    } else {
        // TODO: change python code
        throw new Exception("FilePathsDataReader: next() this should not happen")
    }

    // Check if cur_index has reached the end of data block
    if (curIndex >= curData.length) {
      // check if any more data blocks are left
      if (filePathsReaders.hasNext() == false) {
        curData = null
        curIndex = -1
      } else {
        var found = false
        var flag = true 
        while (flag == true && found == false) {
          if (filePathsReaders.hasNext()) {
            // read the next block and reinitialize the cur_index
            val filePath = filePathsReaders.next()
            curData = FilePathsUtil.read_file_content_as_lines(filePath, s3Region, awsProfile)
            curIndex = 1

            // check if found a valid file
            if (curIndex < curData.length)
              found = true
          } else {
            // invalidate the buffers
            curData = null
            curIndex = -1
            flag = false
          }
        }
      }
    }

    return result
  }
}
